quinn-proto/src/connection/mod.rs

```diff
@@ -773,6 +775,46 @@ impl Connection {
 
         self.app_limited = buf.is_empty() && !congestion_blocked;
 
+        // Send MTU probe if necessary
+        if buf.is_empty() && self.state.is_established() {
+            let space_id = SpaceId::Data;
+            let probe_size = match self
+                .path
+                .mtud
+                .poll_transmit(self.spaces[space_id].next_packet_number)
+            {
+                Some(next_probe_size) => next_probe_size,
+                None => return None,
+            };
+
+            if buf.capacity() < probe_size as usize {
+                buf.reserve(probe_size as usize - buf.capacity());
+            }
+
+            let buf_capacity = probe_size as usize;
+
+            let mut builder = PacketBuilder::new(
+                now,
+                SpaceId::Data,
+                &mut buf,
+                buf_capacity,
+                0,
+                true,
+                self,
+                self.version,
+            )?;
+
+            // We implement MTU probes as ping packets padded up to the probe size
+            buf.write(frame::Type::PING);
+            builder.pad_to(probe_size);
+            builder.finish_and_track(now, self, Some(SentFrames::default()), &mut buf);
```

Comment:
We should set `non_retransmits` in `SentFrames` here to reflect the presence of non-retransmittable ACK-eliciting frames, similar to how other PINGs are handled:

https://github.com/quinn-rs/quinn/blob/d31597f181ad1357761c0179792e09a7803ba46e/quinn-proto/src/connection/mod.rs#L2729-L2735

I don't think any existing logic cares either way, but it might be a landmine in the future.